Split eventsourced remember entity updates into batched StateChange events based on `max-updates-per-write` config

nathanmbrown opened the pull request:
Resolves #32475

lightbend-cla-validator commented:
At least one commit author (nbrown@ureason.com) is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one commit author (nbrown@ureason.com) is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

nathanmbrown committed:
fix: Split remember entity updates into batched StateChange events #32475

* started and stopped entity ids split into separate events containing batches based on the max-updates-per-write config value.

lightbend-cla-validator commented:
Hi @nathanmbrown,

Thank you for your contribution! We really value the time you've taken to put this together.

We see that you have signed the Lightbend Contributors License Agreement before, however, the CLA has changed since you last signed it.
Please review the new CLA and sign it before we proceed with reviewing this pull request:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

nathanmbrown commented:
CLA signed. The original commit had an incorrect old email address on it (nbrown@ureason.com), I have fixed that by pushing a corrected commit.

patriknw reviewed (approved):
LGTM, good catch and thanks for fixing

johanandren reviewed (approved):
LGTM

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntitiesShardStore.scala:
```diff
@@ -93,30 +106,21 @@ private[akka] final class EventSourcedRememberEntitiesShardStore(
       val events =
-        (if (started.nonEmpty) EntitiesStarted(started) :: Nil else Nil) :::
-        (if (stopped.nonEmpty) EntitiesStopped(stopped) :: Nil else Nil)
+        (if (started.nonEmpty) createEntityEvents(started, EntitiesStarted, maxUpdatesPerWrite) else Nil) :::
+        (if (stopped.nonEmpty) createEntityEvents(stopped, EntitiesStopped, maxUpdatesPerWrite) else Nil)
```
Oh, one small thing for Scala 3 compatibility:
```suggestion
        (if (started.nonEmpty) createEntityEvents(started, EntitiesStarted.apply _, maxUpdatesPerWrite) else Nil) :::
        (if (stopped.nonEmpty) createEntityEvents(stopped, EntitiesStopped.apply _, maxUpdatesPerWrite) else Nil)
```

nathanmbrown replied:
Thanks, I was wondering why the Scala 3 compatibility wasn’t passing!

nathanmbrown committed:
Scala 3 compatibility fix

Pass explicit function references to the createEntityEvents function to ensure Scala 3 compatibility

Co-authored-by: Johan Andrén <johan@markatta.com>

johanandren commented:
Sorry, one more place with the same thing, in the test

nathanmbrown committed:
Scala 3 compatibility 

Also apply the previous fix to the tests
